library/entropy_poll.c

```diff
@@ -115,6 +115,42 @@ static int getrandom_wrapper( void *buf, size_t buflen, unsigned int flags )
 #endif /* SYS_getrandom */
 #endif /* __linux__ || __midipix__ */
 
+/*
+ * Some BSD systems provide KERN_ARND.
+ * This is equivalent to reading from /dev/urandom, only it doesn't require an
+ * open file descriptor, and provides up to 256 bytes per call (basically the
+ * same as getentropy(), but with a longer history).
```

Comment:
I've discussed this with other projects and the response has been mixed. Certain libraries expressed a preference towards using a system call because it's harder to reconstruct the RNG state than it is for `arc4random_buf` where the state lives in libc and is per-process. arc4random obtains its initial state using this API, and is faster,  but it does have a somewhat awkward name, and is still using RC4 on some modern operating systems. If this was simply about providing an endless stream of random numbers to applications, I'd use arc4random without question, but since this is an "entropy" API that is intended for seeding other RNGs, it seemed more prudent to use a system call.

Using the sysctl will be safe on antique releases and using it doesn't involve having to check for the presence of a function in libc, just the presence of a \#define (the `sysctl` API is ancient BSD stuff). OpenSSL has an explicit policy of supporting OS versions that even we don't support, so they had a preference for this API.

Speaking from the NetBSD side, I'd like to see more software using `arc4random_buf` (other developers and users may disagree with me), but it's up to you which you prefer to use in mbedtls.

Reply:
I'm starting to wonder if a three-layer approach would make sense:

1. If we're on a set of known-good OS and versions use `arc4random_buf`.
2. Otherwise use `KERN_ARND` if available.
3. Otherwise fall back to `open()`ing `/dev/urandom`.

_If_ we go for this approach, we could begin by finalizing and merging this PR, and then have another one adding the arc4random layer.

The question is: what platforms would benefit from the `KERN_ARND` layer in such an approach?

- FreeBSD 11 still uses ARC4 for `arc4random` and has `KERN_ARND` but it will be EOL in a bit more than a year (end of September 2021).
- Dragonfly BSD looks like it's still using ARC4 for `arc4random`, according to [this man page](https://man.dragonflybsd.org/?command=arc4random&section=ANY), but doesn't have `KERN_ARND` so can't benefit from this layer anyway. 

So maybe three layers are not necessary, and just having `arc4random()` when it's known to be safe, and `/dev/urandom` as a fallback would be enough.

Wdyt?

Reply:
I'm not strongly opinionated about which approach gets chosen — I really just want a better default choice than reading from /dev/urandom for NetBSD. Other than obvious performance differences, the only meaningful difference on all supported versions of NetBSD is that the application can't access the state of the KERN_ARND RNG.

If you're happy using `arc4radom_buf` in an entropy function, I will submit a new PR after this one is closed or merged.

Reply:
Ok, so there is a minor security advantage to the `sysctl` approach: if the arc4random state is compromised by a memory read vulnerability before it's used to seed the entropy for mbedtls, that compromises the mbedtls entropy. (If the arc4random state is compromised later, that shouldn't affect the mbedtls entropy since modern arc4random has backtracking resistance.) It's not such a big advantage that I'd reject arc4random, which is simpler, but since the sysctl code is there, I'm happy to use it.

Regarding OS version support, we don't have a well-defined policy, but we do try to accommodate people who run Mbed TLS on “oddball” platforms, including out-of-support platforms. But we can only promise best effort, not results, when it comes to platforms for which we don't have technical knowledge on the team and a CI system in place.